hs/src/Reach/Optimize.hs

```diff
@@ -786,18 +881,30 @@ instance Optimize DLStmt where
       cr <- asks eConstR
       let f = Just . (+) 1 . fromMaybe 0
       liftIO $ modifyIORef cr $ M.alter f v
-    DL_LocalIf _ _ t f -> gcs t >> gcs f
+    DL_LocalIf _ _ _ t f -> gcs t >> gcs f
     DL_LocalSwitch _ _ csm -> gcsSwitch csm
     DL_ArrayMap _ _ _ _ _ f -> gcs f
     DL_ArrayReduce _ _ _ _ _ _ _ f -> gcs f
     DL_MapReduce _ _ _ _ _ _ _ f -> gcs f
     DL_Only _ _ l -> gcs l
-    DL_LocalDo _ t -> gcs t
+    DL_LocalDo _ _ t -> gcs t
+
+-- Track that we are attempting to lift this assignment
+preFloat :: DLStmt -> App ()
+preFloat = \case
+  DL_LocalIf _ (Just dv) _ _ _ -> do
+    eVarsSetR <- asks eVarsSet
+    eVarsSet <- (liftIO . readIORef) eVarsSetR
+    liftIO $ modifyIORef' eVarsSetR $
+      M.insert dv $ maybe (1) succ $ M.lookup dv eVarsSet
```

Comment:
It should always increment even if it is already set because we're tracking how many `if`s we ought to float in order to drop `DL_Var`. If there's any impure `if`s, `dv` will be associated with a non-zero number and we'll know to keep `DL_Var`